### PR TITLE
Add P-256 signatures/key exchange

### PR DESCRIPTION
**Pull request by l1h3r**
# Description of change

Adds P-256 signatures and key exchange to use with JSON web tokens.

## Type of change

- [x] Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

RFC test vectors

## Change checklist

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by JelleMillenaar** [comment]
@felsweg-iota Can you review?

**Review by felsweg-iota — APPROVED** [review]
Looks good to me. Fix merge conflicts.

File: src/signatures/p256.rs [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+    }
+
+    /// Return the [PublicKey] which corresponds to this [SecretKey].
+    pub fn public_key(&self) -> PublicKey {
+        PublicKey(self.0.public_key())
+    }
+
+    /// Sign the given `message` and return a digital [signature][Signature].
+    pub fn sign(&self, message: &[u8]) -> crate::Result<Signature> {
+        self.try_sign(message)
```
Comment (semenov-vladyslav) [inline src/signatures/p256.rs:137]: What to do if `try_sign` fails?

File: Cargo.toml [hunk 2]
```diff
@@ -123,6 +124,18 @@ chacha20poly1305 = { version = "0.7.1", optional = true }
   optional = true
   features = [ "derive" ]
 
+  [dependencies.p256-crate]
```
Comment (semenov-vladyslav) [inline Cargo.toml:127]: Didn't notice when dependencies switch to this format. Hope, no dependencies were lost

File: src/error.rs [hunk 3]
```diff
@@ -14,9 +14,11 @@ pub enum Error {
         needs: usize,
         has: usize,
     },
-    ///  Cipher Error
+    /// Cipher Error
     CipherError { alg: &'static str },
-    /// Convertion Error
+    /// Signature Error
+    SignatureError { alg: &'static str },
```
Comment (semenov-vladyslav) [inline src/error.rs:20]: Some of these changes were proposed in other PRs

File: src/signatures/p256.rs [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+        p256_crate::ecdsa::VerifyingKey::from(&self.0).verify(message, &signature.0)
+    }
+}
+
+/// A P-256 Secret Key
+pub struct SecretKey(p256_crate::SecretKey);
+
+impl SecretKey {
+    /// Generate a new random [SecretKey].
+    #[cfg(feature = "random")]
```
Comment (semenov-vladyslav) [inline src/signatures/p256.rs:106]: Should there also be `#[cfg_attr(docsrs, doc(cfg(feature = "random")))]`?

**Review by semenov-vladyslav — CHANGES_REQUESTED** [review]
Looks good, same style and secret key leakage via local variable notes.

**Comment by Wollac** [comment]
Seems like this PR is no longer needed. Please just re-open if it still is.

**Pull request closed without merging**